sound/soc/sof/topology.c

```diff
@@ -1385,67 +1375,167 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 	kfree(tone);
 	return ret;
 }
-/*
- * Effect Topology. Only IIR equalizer is supported at this moment.
- * TODO: Need to add also FIR support and have a way to add other
- * effects and enhancements.
- */
 
-static int sof_widget_load_effect(struct snd_soc_component *scomp, int index,
-				  struct snd_sof_widget *swidget,
-				  struct snd_soc_tplg_dapm_widget *tw,
-				  struct sof_ipc_comp_reply *r)
+static int sof_effect_fir_load(struct snd_soc_component *scomp, int index,
+			       struct snd_sof_widget *swidget,
+			       struct snd_soc_tplg_dapm_widget *tw,
+			       struct sof_ipc_comp_reply *r)
+
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_soc_tplg_private *private = &tw->priv;
-	struct sof_ipc_comp_eq_iir *eq;
+	struct sof_ipc_comp_eq_fir *fir;
+	size_t ipc_size = 0, fir_data_size = 0;
+	void *fir_data_p;
 	int ret;
 
-	eq = kzalloc(sizeof(*eq), GFP_KERNEL);
-	if (!eq)
+	fir_data_size = le32_to_cpu(private->size) - SOF_EFFECT_DATA_SIZE;
+	ipc_size = sizeof(struct sof_ipc_comp_eq_fir) + fir_data_size;
+	fir_data_p = private->data + SOF_EFFECT_DATA_SIZE;
+
+	fir = kzalloc(ipc_size, GFP_KERNEL);
+	if (!fir)
 		return -ENOMEM;
 
-        /* configure IIR EQ IPC message */
-	eq->comp.hdr.size = sizeof(*eq);
-	eq->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
-	eq->comp.id = swidget->comp_id;
-	eq->comp.pipeline_id = index;
+	/* configure mixer IPC message */
+	fir->comp.hdr.size = ipc_size;
+	fir->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
+	fir->comp.id = swidget->comp_id;
+	fir->comp.type = SOF_COMP_EQ_FIR;
+	fir->comp.pipeline_id = index;
 
-	eq->comp.type = SOF_COMP_EQ_IIR;
-	ret = sof_parse_tokens(scomp, eq, eq_iir_tokens,
-			       ARRAY_SIZE(eq_iir_tokens), private->array,
-			       le32_to_cpu(private->size));
-	if (ret) {
-		dev_err(sdev->dev, "error: parse EQ tokens failed %d\n",
-			private->size);
-		goto err;
-	}
+	memcpy(&fir->data, fir_data_p, fir_data_size);
+
+	fir->size = fir_data_size;
 
-	ret = sof_parse_tokens(scomp, &eq->config, comp_tokens,
+	ret = sof_parse_tokens(scomp, &fir->config, comp_tokens,
```

Comment:
can this be moved to load_effect instead of duplicating it for load_iir and load_fir?

Reply:
Yes will try to do it.

Reply:
now thinking of this, the ipc struct (including config) is created only in the specialized effect loading. I should create separate config then in effect_load and send it as parameter to special effect loading. IMHO actually does not help too much and makes things little bit more confusing (also need to take care of extra error cases).